File: solutions/r/bob/1/bob.R
bob <- function(input) {
  # 1. Trim leading and trailing whitespace characters
  trimmed <- gsub("^[[:space:]]+|[[:space:]]+$", "", input)
  
  # 2. Check if the input is silent (empty after trimming whitespace)
  is_silence <- nchar(trimmed) == 0
  
  if (is_silence) {
    return("Fine. Be that way!")
  }
  
  # 3. Check if the input ends with a question mark
  is_question <- grepl("\\?$", trimmed)
  
  # 4. Check if the input is yelled (has letters, and all letters are uppercase)
  # grepl("[[:alpha:]]", trimmed) ensures there is at least one letter
  is_yelling <- (trimmed == toupper(trimmed)) && grepl("[[:alpha:]]", trimmed)
  
  # 5. Determine Bob's response based on the conditions
  if (is_yelling && is_question) {
    return("Calm down, I know what I'm doing!")
  } else if (is_yelling) {
    return("Whoa, chill out!")
  } else if (is_question) {
    return("Sure.")
  } else {
    return("Whatever.")
  }
}

File: solutions/euphoria/bob/1/bob.ex
include std/text.e

public function hey(sequence s)
  s = trim(s)
  if equal(s, "") then
    return "Fine. Be that way!"
  end if
  atom isQuestion = s[$] = '?'
  atom isYelling = equal(s, upper(s)) and not equal(lower(s), upper(s))
  if isQuestion and isYelling then
    return "Calm down, I know what I'm doing!"
  elsif isQuestion then
    return "Sure."
  elsif isYelling then
    return "Whoa, chill out!"
  else
    return "Whatever."
  end if
end function